Sprint 07 Final

Heamex opened the pull request.

Heamex committed:
Внедрены тесты

Heamex committed:
Начало работы по рефакторингу проекта

Создал презентер и перенёс в него пару переменных и несколько функций из MovieQuizViewController

Heamex committed:
Кнопки из VC перенёс в Presenter

Heamex committed:
Рефакторинг, шаг 3

Продолжаю переносить функциональность в презентер

Heamex committed:
Рефакторинг ч. 4

Перенёс ещё больше функционала в презентер

Heamex committed:
Рефакторинг окончен

На данном этапе закончил рефакторинг.
Теперь во MovieQuizViewController только представление и отсутствует бизнес-логика.

Heamex committed:
Рефакторинг шаг 4

Heamex committed:
Проект готов

Настроил UI и Unit тесты, закончил рефакторинг.
Приложение успешно проходит все тесты, показывает вопросы, даёт ответы, ведёт статистику.

tsaidov7 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,175 +1,103 @@
+	@IBOutlet var textLabel: UILabel!
+	@IBOutlet var counterLabel: UILabel!
+	@IBOutlet var noButton: UIButton! // тут оутлеты
+	@IBOutlet var yesButton: UIButton! // на две кнопки
```
private
Указывай private для того, чтобы спрятать детали реализации конкретной части функциональности,
когда они используются только внутри области объявления

tsaidov7 commented on Services/MovieQuizPresenter.swift:
```diff
@@ -0,0 +1,136 @@
+		currentQuestion = question
+		let viewModel = convert(model: question)
+		DispatchQueue.main.async { [weak viewController] in
+			viewController?.showQuiz(quiz: viewModel)
```
Ты и так слабо на контроллер ссылаешься, зачем еще раз устанавливать weak ссылку?)

tsaidov7 commented on MovieQuiz/Presentation/MovieQuizViewController.swift:
```diff
@@ -1,175 +1,103 @@
 
-final class MovieQuizViewController: UIViewController, QuestionFactoryDelegate, AlertPresenterDelegate {
-	
+final class MovieQuizViewController: UIViewController, MovieQuizViewControllerProtocol {
```
Предпочитаю конформ к протоколу выносить в расширение класса для ускорения работы компилятора. Это оптимизация его работы.
https://habr.com/ru/post/673400/

Heamex committed:
Ещё небольшие правочки

Приватизировал все UI элементы в MovieQuizViewController.

Также в MovieQuizViewController, и MovieQuizPresenter вынес конформ протокола extantion для ускорения компиляции приложения
